diff -ru SDL.orig/src/render/SDL_render.c SDL/src/render/SDL_render.c
--- SDL.orig/src/render/SDL_render.c	2018-10-31 16:07:22.000000000 +0100
+++ SDL/src/render/SDL_render.c	2018-12-02 11:48:59.000000000 +0100
@@ -2279,15 +2279,13 @@
 SDL_BlendFactor
 SDL_GetBlendModeDstAlphaFactor(SDL_BlendMode blendMode)
 {
-    blendMode = SDL_GetLongBlendMode(blendMode);
-    return (SDL_BlendFactor)(((Uint32)blendMode >> 24) & 0xF);
+    return SDL_GetBlendModeDstColorFactor(blendMode);
 }
 
 SDL_BlendOperation
 SDL_GetBlendModeAlphaOperation(SDL_BlendMode blendMode)
 {
-    blendMode = SDL_GetLongBlendMode(blendMode);
-    return (SDL_BlendOperation)(((Uint32)blendMode >> 16) & 0xF);
+    return SDL_GetBlendModeColorOperation(blendMode);
 }
 
 /* vi: set ts=4 sw=4 expandtab: */
